#166284571 Chore restructure the files directories

despeauxz opened the pull request:
#### What does this PR do?
This PR restructures the entire project directories
#### Description of Task to be completed?
- Have a root folder named `server`
#### How should this be manually tested?
Check the repo [https://github.com/andela/kingsmen-ah-backend](https://github.com/andela/kingsmen-ah-backend)
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#166284571](https://www.pivotaltracker.com/story/show/166284571)

#### Screenshots (if appropriate)
N/A
#### Questions:
N/A

despeauxz committed:
chore-[#166284571]: restructure the files dir

- use single root directory(server)

[Delivers #166284571]

fidelisojeah reviewed (changes requested):
Wait you believe that all these files you have deleted are useless?

despeauxz committed:
chore-[166284571]: server setup

- server setup

[Delivers #166284571]

despeauxz committed:
chore-[166284571]: setup postgres

- setupt postgres engine
- added the .env file to gitignore

[Delivers #166284571]

fidelisojeah committed:
Merge branch 'develop' into 166284571-ch-refactor-files-structure

fidelisojeah committed:
chore-[166284571]: refactor file structure

fidelisojeah commented on server/config/database_config.js:
```diff
@@ -0,0 +1,26 @@
+require('dotenv').config();
```
?

fidelisojeah commented on server/models/index.js:
```diff
@@ -0,0 +1,42 @@
+const path = require('path');
```
?

despeauxz committed:
chore-[166284571]: postgres setup with passport

- re-create existing routes
- setup passport

[Delivers 166284571]

despeauxz committed:
Merge branch '166284571-ch-refactor-files-structure' of https://github.com/andela/kingsmen-ah-backend into 166284571-ch-refactor-files-structure

fidelisojeah commented on server/routes/api/users.js:
```diff
@@ -0,0 +1,86 @@
+    .catch(next);
+});
+
+router.post('/users/login', (req, res, next) => {
```
move logic to `controllers`

fidelisojeah commented on server/routes/api/users.js:
```diff
@@ -0,0 +1,86 @@
+  })(req, res, next);
+});
+
+router.post('/users', (req, res, next) => {
```
move logic to `controllers`

fidelisojeah reviewed (commented):
Also update the readme with current setup (or consider creating a wiki) to help teammates and everyone new in the team to set up properly

despeauxz committed:
chore-[166284571]: refactor file structure

- refactor file structure
- use mvc pattern

[Delivers #166284571]

despeauxz committed:
chore-[166284571]: refactor file structure

- refactor file structure
- use mvc pattern

[Delivers #166284571]

despeauxz committed:
chore-[166284571]: refctore codebase

- refactor codebase
- added extra script commands
- wrapped my methinds in try-catch block

[Delivers #166284571]

fidelisojeah commented on package.json:
```diff
@@ -18,24 +24,20 @@
     "@babel/preset-env": "^7.4.5",
     "@babel/register": "^7.4.4",
+    "@babel/runtime": "^7.4.5",
+    "bcryptjs": "^2.4.3",
```
please use `bcrypt` rather than `bcryptjs`

fidelisojeah commented on server/config/database_config.js:
```diff
@@ -0,0 +1,28 @@
+    host: process.env.DB_HOST,
+    dialect: 'postgres'
+  },
+  test: {
```
consider using `sqlite` for testing

fidelisojeah commented on server/config/database_config.js:
```diff
@@ -0,0 +1,28 @@
+
+config();
+
+module.exports = {
```
this can all be broken down into
```javascript
export const development = {};
export const test = {};
export const production = {};
```
??

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -0,0 +1,124 @@
+  * @static
+  */
+  static async login(req, res, next) {
+    if (!req.body.email) {
```
consider using/creating some sort of validator for this.

would be nice/it is possible to have these multiple validations done at once (validate email, password etc)

fidelisojeah commented on server/app.js:
```diff
@@ -0,0 +1,26 @@
+}));
+
+app.use('/', apiRoutes);
+
```
considering that there was an error handler in the original file,

```javascript
/// catch 404 and forward to error handler	
app.use((req, res, next) => {	
    const err = new Error("Not Found");	
    err.status = 404;	
    next(err);	
});	

 /// error handlers	

 // development error handler	
// will print stacktrace	
if (!isProduction) {	
    app.use((err, req, res, next) => {	
        console.log(err.stack);	

         res.status(err.status || 500);	

         res.json({	
            errors: {	
                message: err.message,	
                error: err	
            }	
        });	
    });	
}	

 // production error handler	
// no stacktraces leaked to user	
app.use((err, req, res, next) => {	
    res.status(err.status || 500);	
    res.json({	
        errors: {	
            message: err.message,	
            error: {}	
        }	
    });	
});
```

try improving on it or just put it back

fidelisojeah committed:
chore-(166284571) :  fix merge conflicts

despeauxz committed:
chore-[166284571]: refactor file structure

- improve error handler
- implement route validation

[Finishes #166284571]

despeauxz committed:
chore-[166284571]: refactor file structure

- improve error handler
- implement route validation

[Finishes #166284571]

despeauxz committed:
chore-[166284571]: refactor file structure

- improve error handler
- implement route validation

[Finishes #166284571]

fidelisojeah commented on server/validations/users.js:
```diff
@@ -0,0 +1,82 @@
+import { check } from 'express-validator/check';
```
This is not exactly what I meant.

I meant you should use some sort of schema validation which can happen in each controller.

While this may work, this seems very complicated to set up. 

consider `yup` or `joi` or writing a validator which can be easily set up/used.

If this works well for you though let me know

fidelisojeah commented on server/middlewares/ValidationHandler.js:
```diff
@@ -0,0 +1,59 @@
+    if (!errors.isEmpty()) {
+      const mappedErrors = errors.mapped();
+
+      return res.status(422).json({
```
any reason you are using the `422` error code here?

validation errors are usually `400` (Bad requests)

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -0,0 +1,126 @@
+      const user = await User.findByPk(req.payload.id);
+
+      if (!user) {
+        return res.sendStatus(401);
```
Are you just using random error codes?

despeauxz committed:
chore-[166284571]: refactor file structure

- improve error handler
- implement route validation

[Finishes #166284571]

despeauxz committed:
chore-[166284571]: refactor file structure

- improve error handler
- implement route validation

[Finishes #166284571]

fidelisojeah commented on server/config/database_config.js:
```diff
@@ -0,0 +1,32 @@
+};
+
+export const test = {
+  username: process.env.DB_USERNAME || 'postgres',
```
remove this

The `sqlite` should not need `username` `password` etc

`dialect` and `storage` should be enough

despeauxz committed:
Merge branch '166284571-ch-refactor-files-structure' of https://github.com/andela/kingsmen-ah-backend into 166284571-ch-refactor-files-structure

despeauxz committed:
Merge branch '166284571-ch-refactor-files-structure' of https://github.com/andela/kingsmen-ah-backend into 166284571-ch-refactor-files-structure

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -0,0 +1,148 @@
+        return res.status(400).json({ status: 400, error: errorValue });
+      }
+      const { username, email, password } = req.body;
+      const validUsername = await User.findOne({
```
instead of doing this here,I guess you could use your `findOrCreate` with `email or username`.

additionally, you could also just use the `User.create` then check the error it revels in the catch block and (if it a unique violation) return the `email already taken` or `username already taken` error (this would usually be in `err.error`) or if it's not a unique violation then you `next(err)`

fidelisojeah reviewed (commented):
Good job.

Just a few changes and also if you would update the readme on how to setup

lastly, add `*.sqlite` to your `gitignore`

fidelisojeah commented on server/config/database_config.js:
```diff
@@ -0,0 +1,28 @@
+  storage: path.join(__dirname, '..', 'database_test.sqlite'),
+};
+
+export const production = {
```
change this to:
```javascript
export const production = {
    use_env_variable: process.env.DATABASE_URL,
    dialect: 'postgres',
    dialectOptions: {
      ssl: true,
    },
    logging: false,
  }
```

this difference here is you had `production` as a member of `const production`

despeauxz committed:
Merge branch 'develop' of https://github.com/andela/kingsmen-ah-backend into 166284571-ch-refactor-files-structure

despeauxz committed:
Merge branch '166284571-ch-refactor-files-structure' of https://github.com/andela/kingsmen-ah-backend into 166284571-ch-refactor-files-structure

despeauxz committed:
Merge branch '166284571-ch-refactor-files-structure' of https://github.com/andela/kingsmen-ah-backend into 166284571-ch-refactor-files-structure

fidelisojeah commented on .sequelizerc:
```diff
@@ -0,0 +1,8 @@
+    "models-path": path.resolve('./server/models'),
+    "seeders-path": path.resolve('./server/seeders'),
+    "migrations-path": path.resolve('./server/migrations')
+};
```
new line here

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -0,0 +1,149 @@
+      }
+      const { username, email, password } = req.body;
+      const [user, created] = await User.findOrCreate({
+        where: { email: { [Op.iLike]: email } },
```
`[Op.iLike]` you likely want to use `[Op.eq]` here.
However, I suggest you still use just `User.create` rather than `User.findOrCreate` here then the catch block would look like:

```javascript
if (err.errors && err.errors[0].type === 'unique violation') {
        return res.status(400).json({
          status: 400,
          errors: err.errors.reduce((result, currentValue) => {
            if (result.type === 'unique violation'){
              result[currentValue.path] = `${currentValue.path} has already been taken`;
            } else if (currentValue.path) {
              result[currentValue.path] = currentValue.message;
            } else {
              result['global'] = currentValue.message;
            }
            return result;
          }, {})
        });
      }
```

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -0,0 +1,149 @@
+  */
+  static async create(req, res, next) {
+    try {
+      const { error } = validateSignup(req.body);
```
consider modifying this.

```javascript
try{
  const userDetails = await validateSignup(req.body);

  const user = await User.create(userDetails);

  return res.status(201).send({ status: 'success', message: 'User created successfully' });
catch (err) {
      if (err.isJoi && err.name === 'ValidationError') {
        return res.status(400).json({
          status: 400,
          errors: err.details.reduce((result, currentValue) => {
            if (!Object.hasOwnProperty.call(result, currentValue.context.key)) {
              result[currentValue.context.key] = currentValue.message;
            }
            return result;
          }, {})
        });
      }
...
```

You should do something similar everywhere else

fidelisojeah commented on server/validations/auth.js:
```diff
@@ -0,0 +1,43 @@
+  return Joi.validate(user, schema);
+};
+
+export const validateSignup = (user) => {
```
consider defining the schema with:

``` javascript
const schema = {
    username: Joi.string().trim().lowercase().min(4).max(100)
      .required()
      .alphanum(),
    email: Joi.string().trim().lowercase().min(5).max(255)
      .email()
      .required(),
    password: Joi.string().trim().min(5).max(255)
      .required()
  };

return Joi.validate(user, schema, {
    language: {
      key: '{{key}} '
    },
    abortEarly: false,
    stripUnknown: true
  });
```
You would do something similar for other schemas .

the 
```javascript
language: {
      key: '{{key}} '
}
```

also notice the `.lowercase()` included

means you don't need to remove the quotes around the key manually as you have been doing in the code.
Also the ` abortEarly: false` ensures that all fields are validated and it wouldn't just be exited on first error

fidelisojeah commented on server/models/index.js:
```diff
@@ -0,0 +1,44 @@
+const db = {};
+
+let sequelize;
+if (config.use_env_variable) {
```
Modify this:

```javascript
if (config.use_env_variable) {
  sequelize = new Sequelize(process.env[config.use_env_variable], config);
} else {
  const {
    database,
    username,
    password,
    ...otherConfig
  } = config;
  sequelize = new Sequelize(database, username, password, {
    ...otherConfig
  });
}
```

notice I removed `operatorsAliases` (it now defaults to false: which is what we want). Since you are already using things like `[Op.iLike]`

fidelisojeah reviewed (changes requested):
Good job generally.

Just a few more changes

despeauxz committed:
Merge branch '166284571-ch-refactor-files-structure' of https://github.com/andela/kingsmen-ah-backend into 166284571-ch-refactor-files-structure

despeauxz committed:
Merge branch '166284571-ch-refactor-files-structure' of https://github.com/andela/kingsmen-ah-backend into 166284571-ch-refactor-files-structure

despeauxz committed:
Merge branch '166284571-ch-refactor-files-structure' of https://github.com/andela/kingsmen-ah-backend into 166284571-ch-refactor-files-structure
